File: src/h5create.m
##
##    Copyright (C) 2012 Tom Mullins
##    Copyright (C) 2015 Tom Mullins, Thorsten Liebig, Anton Starikov, Stefan Großhauser
##    Copyright (C) 2008-2013 Andrew Collette
##
##    This file is part of hdf5oct.
##
##    hdf5oct is free software: you can redistribute it and/or modify
##    it under the terms of the GNU Lesser General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    hdf5oct is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU Lesser General Public License for more details.
##
##    You should have received a copy of the GNU Lesser General Public License
##    along with hdf5oct.  If not, see <http://www.gnu.org/licenses/>.
##

# -*- texinfo -*-
# @deftypefn {Function File} { } h5create (@var{filename}, @var{dsetname}, @var{size}, @var{key}, @var{val},...)
#
# Create a dataset with name @var{dsetname} and size @var{size}
# in the HDF5 file specified by @var{filename}.  Intermediate groups
# are created as necessary.
#
# The vector @var{size} may contain one or several Inf (or
# equivalently: zero) values.
# This will lead to unlimited maximum extent of the dataset in the
# respective dimensions and 0 initial extent.
# Note that any dataset with at least one unlimited dimension must be chunked and
# it is generally recommended for large datasets.
#
# The list of @var{key}, @var{val} arguments allows to specify
# certain properties of the dataset. Allowed settings are:
#
# @table @asis
# @item @option{Datatype}
# one of the strings: 
# @samp{double}(default) | @samp{single} | @samp{uint64} | @samp{uint32} | 
# @samp{uint16} | @samp{uint8} | @samp{int64} | @samp{int32} | @samp{int16} | 
# @samp{int8} | 'string'
# @item @option{ChunkSize}
# The value may be either a vector specifying the chunk size,
# or an empty vector [], which means no chunking (this is the default),
# or the string @samp{auto} which makes the library choose automatically
# an appropriate chunk size, as best as it can. Note that the @samp{auto}
# setting is not @sc{matlab} compatible.
# @end table
#
# @seealso{h5write}
# @end deftypefn


function h5create(filename,location,sz,varargin)

# check number and types of arguments
if ((nargin < 3))
    error("h5create: you must supply at least 3 arguments");
endif
if (!ischar(filename))
  error("h5create: 1st argument must be a string holding the hdf5 file name");
endif
create_file = 1;
if isfile(filename)
  # check valid hdf5
  create_file = 0;
endif
if (!ischar(location))
  error("h5create: 2nd argument must be a string holding the dataset location");
endif
if !isreal(sz) || !isvector(sz),
  error("h5create: 3rd argument must be a dimension vector");
endif
sz = sz(:);
# sz elements must be index or Inf
for i=1:length(sz),
  if !(isindex(sz(i)) || sz(i)==inf)
    error("h5create: 3rd argument elements must be valid index values or inf");
  endif
end

## check options
[reg, datatype, chunksize, fillvalue] = parseparams (varargin, ...
  'Datatype', 'double',...
  'ChunkSize',[],...
  'FillValue',0);

# check datatype
if !(strcmp(datatype,'double') || ...
  strcmp(datatype,'single') || ...
  strcmp(datatype,'uint64') || ...
  strcmp(datatype,'int64') || ...
  strcmp(datatype,'uint32') || ...
  strcmp(datatype,'int32') || ...
  strcmp(datatype,'uint16') || ...
  strcmp(datatype,'int16') || ...
  strcmp(datatype,'uint8') || ...
  strcmp(datatype,'int8') || ...
  strcmp(datatype,'string'))
  error("h5create: invalid 'Datatype'");
endif
# check
if !isindex(sz) && isempty(chunksize)
  error("h5create: chunksize must be defined for inf size datasets");
endif
if !isempty(chunksize) 
  if size(chunksize(:))!=size(sz),
    error("h5create: chunksize is different size from 3rd argument");
  end
  if !isindex(chunksize),
    error("h5create: invalid chunksize");
  end
  chunksize = chunksize(:);
end

if size(sz,1)==1, # convert [n] to [1xn]
  sz = [sz; 1];
  if !isempty(chunksize)
    chunksize = [chunksize; 1];
  end
end

__h5create__(filename,create_file,location,sz,datatype,chunksize,fillvalue);

File: src/__h5create__.cc
/*
 *
 *    Copyright (C) 2012 Tom Mullins
 *    Copyright (C) 2015 Tom Mullins, Thorsten Liebig, Anton Starikov, Stefan Großhauser
 *    Copyright (C) 2008-2013 Andrew Collette
 *
 *    This file is part of hdf5oct.
 *
 *    hdf5oct is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU Lesser General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    hdf5oct is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public License
 *    along with hdf5oct.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
// integrated into the GNU Octave build
#include "oct.h"
#include "lo-ieee.h"
#else
// as a package
#include <octave/oct.h>
#include <octave/lo-ieee.h>
#include <octave/file-stat.h>
#include <octave/ls-hdf5.h>
#include <octave/oct-stream.h>
#endif

#include <cstdlib>
#include <cerrno>
#include <iostream>
#include <algorithm>
#include <string>

#include "hdf5oct.h"

using namespace std;
namespace H5 = HighFive;
namespace h5o = hdf5oct;

// __h5create__(fname,create_file,loc,sz,datatype,chunksize,fillvalue)
DEFUN_DLD(__h5create__, args, ,"__h5create__: backend for h5create\n\
Users should not use this directly. Use h5create.m instead")
{
    string filename = args(0).string_value();
    bool create_file = args(1).bool_value();
    string location = args(2).string_value();
    uint64NDArray size = args(3).uint64_array_value();
    string datatype = args(4).string_value();
    uint64NDArray chunksize = args(5).uint64_array_value();
    int fillvalue = args(6).int_value();

    bool has_unlimited = !chunksize.isempty();

    try {

        // H5::Exception::dontPrint();

        //open the hdf5 file, create it if it does not exist
        H5::File file(filename, create_file ? H5::File::Create : H5::File::ReadWrite);

         if (h5o::locationExists(file,location)) {
               error("h5create: location '%s' already exists",location.c_str());
               return octave_value();
            }        

         H5::DataSpace fspace = H5::DataSpace::Scalar();
         size_t ndim = size.numel(); 
         bool is_scalar = (ndim==1 && size_t(size(0))==1) ||
                           (ndim==2 && size_t(size(0))==1 & size_t(size(1))==1);
         if (!is_scalar) {
            vector<size_t> dims(ndim), maxdims(ndim);
            for(int i=0; i<ndim; i++) {
               bool b = size_t(size(i)) == H5::DataSpace::UNLIMITED;
               dims[ndim-1-i] = b ? 0 : size_t(size(i));
               maxdims[ndim-1-i] = size_t(size(i));
            }
            fspace = H5::DataSpace(dims, maxdims);
         }

        /*
        * Modify dataset creation properties, i.e. enable chunking.
        */
        H5::DataSetCreateProps dscp;

      if (has_unlimited) { 
         vector<hsize_t> chunk_dims(ndim);  
         for(int i=0; i<ndim; i++) chunk_dims[ndim-1-i] = chunksize(i);
            dscp.add(H5::Chunking(chunk_dims));         
      }
 
        /*
        * Set fill value for the dataset
        */
        // cparms.setFillValue( PredType::NATIVE_INT, &fill_val);
 
        /*
        * Create a new dataset within the file using dscp
        * creation properties.
        */
       file.createDataSet(location,fspace,h5o::h5type_from_spec(datatype),dscp);
    }
    catch (const H5::Exception& err) {
        // catch and print any HDF5 error
        error("%s",err.what());
    }

  
  return octave_value ();

}


